[Theme] Fix a few unthemable elements

pdf opened the pull request:
Used palette.textColor to replace hard-coded blacks, should work for
both dark and light themes

Closes #3135.

oliviertassinari commented:
@pdf Thanks for this PR, but let us fix https://github.com/callemall/material-ui/issues/2901 before reviewing it.

pdf commented:
@oliviertassinari sounds like a fair plan.  FYI, I found a bunch more hard-coded values, I think I've got nearly all of them, except for one or two like error colors and media overlays.  Should be pretty close to having themes apply cleanly now, though there may still be some places where the doc default text color was being assumed.

I did also change the `ColorManipulator` API a little, exposing `luminance` and adding an `opacity` param to `lighten` (on that note, any idea why `lighten` forces opacity to 0.15 by default?).  The changes there should be backwards-compatible.

pdf commented:
Rebased against master

oliviertassinari commented on src/avatar.jsx:
```diff
@@ -86,6 +83,10 @@ const Avatar = React.createClass({
     this.setState({muiTheme: newMuiTheme});
   },
 
+  getTheme() {
```
Can we remove the usage of `getTheme` method from this PR? It's always called multiple times inside the same function. That seems inefficien.

pdf replied:
I went with this logic because it is used already in multiple components...

oliviertassinari replied:
> I went with this logic because it is used already in multiple components...

That's true, we should be consistent regarding this point, but that's not the case right now.
@alitaheri @newoga What do you think?

newoga replied:
Yeah I think we should stop using the function. I haven't see the `getTheme` function pattern around in the library much but if it's around, we should move towards not doing it.

Personally, even if we need a function to help with codeflow/logic I'd rather put the function at the top of the module rather as a member of the React component. The more functions we have of components, the more likely someone will read source and abuse them.

Actually, this is one of the reasons why I prefer using functional components when possible over classes because it just takes the choice away.

In addition, this particular issue is why I was hoping we could figure out something like #2509. A lot of times when we do have the theme, we do have to traverse down a lot of nested properties to get what we need. I'm still not sure best approach but I'm not entirely convinced our current one is the best one.

alitaheri replied:
I agree with the removal of `getTheme` it seems really redundant.

>  we do have to traverse down a lot of nested properties to get what we need

It's not us it's javascript :laughing: :laughing: although nested destructuring takes some of the pain away.

newoga replied:
> It's not us it's javascript :laughing: :laughing: although nested destructuring takes some of the pain away

Haha! I agree about regarding the nested destructuring

@pdf @oliviertassinari Let's get rid of the getTheme function, looking again this is fairly simple component, and I think it's fine just breaking it up earlier into a `const` in the render method. Ironically this component could become a functional component once `style-propable` is gone.

oliviertassinari commented on src/avatar.jsx:
```diff
@@ -129,11 +130,11 @@ const Avatar = React.createClass({
         lineHeight: size + 'px',
         fontSize: size / 2 + 4,
-        color: color,
+        color: color || this.getTheme().color,
```
That's introducing a breaking change.
The doc:
![capture d ecran 2016-01-21 a 14 54 17](https://cloud.githubusercontent.com/assets/3165635/12482246/e1f186f0-c04e-11e5-998b-8ae64173ed36.png)
The spec:
![capture d ecran 2016-01-21 a 14 55 00](https://cloud.githubusercontent.com/assets/3165635/12482261/f3ec46e2-c04e-11e5-83d7-1ee314d38b4a.png)

oliviertassinari replied:
Edit, the color is applied twice. Here that seems uselss.

pdf replied:
Where is the color applied twice?  I'll update the theme to match the spec.

oliviertassinari replied:
By bad, it's here for when we use a letter as a child.

oliviertassinari commented on src/avatar.jsx:
```diff
@@ -141,7 +142,7 @@ const Avatar = React.createClass({
 
       const iconElement = icon ? React.cloneElement(icon, {
-        color: color,
+        color: color || this.getTheme().color,
```
See the comment for the color on the `root` element.

oliviertassinari commented on src/ripples/circle-ripple.jsx:
```diff
@@ -25,11 +35,29 @@ const CircleRipple = React.createClass({
     };
   },
 
+  getInitialState() {
```
Can you move this logic to `TouchRipple`? It would be better to keep this component stateless.

pdf replied:
I'll take a look at what's involved, first pass I was just trying to remove all hard-coded values.

oliviertassinari commented on src/utils/color-manipulator.js:
```diff
@@ -97,8 +97,11 @@ export default {
-  // Desaturates rgb and sets opacity to 0.15
-  lighten(color, amount) {
+  // Desaturates rgb and sets opacity (defaults to 0.15)
+  lighten(color, amount, opacity) {
```
What about `lighten(color, amount, opacity = '0.15) {`instead?

pdf replied:
Sure thing, I'll take another pass at these changes on the weekend.

oliviertassinari replied:
awesome, thanks for this PR :+1:.

newoga commented:
@pdf Thanks for your work on this!

pdf commented:
I believe this commit addresses feedback.  It looks like the list of components below currently use the `getTheme()` pattern, should they be updated for consistency?  There's also inconsistency in the use of a `getStyle()` property across various components.  I'd be tempted to look at these things in a separate PR though.

`grep -Irl getTheme() src/`:
```
src/linear-progress.jsx
src/toolbar/toolbar-title.jsx
src/toolbar/toolbar-separator.jsx
src/toolbar/toolbar.jsx
src/slider.jsx
src/raised-button.jsx
src/radio-button.jsx
src/table/table-row.jsx
src/table/table-header.jsx
src/table/table-header-column.jsx
src/table/table.jsx
src/table/table-footer.jsx
src/table/table-row-column.jsx
src/time-picker/time-display.jsx
src/time-picker/clock-pointer.jsx
src/time-picker/clock-number.jsx
src/time-picker/time-picker-dialog.jsx
src/menu/menu.jsx
src/menu/menu-item.jsx
src/menu/link-menu-item.jsx
src/menu/subheader-menu-item.jsx
src/enhanced-switch.jsx
src/checkbox.jsx
src/circular-progress.jsx
src/floating-action-button.jsx
src/refresh-indicator.jsx
src/toggle.jsx
src/date-picker/year-button.jsx
src/date-picker/date-display.jsx
src/date-picker/day-button.jsx
```

oliviertassinari commented on src/lists/list.jsx:
```diff
@@ -101,7 +103,7 @@ const List = React.createClass({
 
       subheader: {
-        color: Typography.textLightBlack,
+        color: theme.subheaderTextColor,
```
Will conflict with https://github.com/callemall/material-ui/pull/2979/files#diff-509c50933d8b7384457d8c9e34a34374L104.

oliviertassinari commented on src/ripples/focus-ripple.jsx:
```diff
@@ -57,6 +80,8 @@ const FocusRipple = React.createClass({
   },
 
   _getRippleElement(props) {
+    const theme = this.state.muiTheme.ripple;
```
It's already available as a property `this.props.muiTheme`. Can you remove the state logic?

pdf replied:
Is that intentional, rather than the standard state method?  Does doing it this way here provide advantage/disadvantage over obtaining it from the context?

oliviertassinari replied:
It's intentional. The simpler our component are, the best it's. Here, we can avoid to add an additionnal state and context property and keep our component stateless.

oliviertassinari commented on src/ripples/touch-ripple.jsx:
```diff
@@ -55,10 +66,19 @@ const TouchRipple = React.createClass({
   },
 
   start(e, isRippleTouchGenerated) {
+    const theme = this.state.muiTheme.ripple;
```
It's already available as a property `this.props.muiTheme`. Can you remove the state logic?

oliviertassinari commented on src/styles/getMuiTheme.js:
```diff
@@ -93,6 +104,9 @@ export default function getMuiTheme(baseTheme, muiTheme) {
       width: spacing.desktopKeylineIncrement * 4,
       color: palette.canvasColor,
     },
+    list: {
```
Will be subheader once https://github.com/callemall/material-ui/pull/2979 is merged.

oliviertassinari commented:
> It looks like the list of components below currently use the getTheme() pattern, should they be updated for consistency?

Yes, we should do it at some point but I think that it's out of the scope of this PR.

pdf commented:
Actioned feedback, rebased against master.  Will update again if you want to merge #2979 first.

oliviertassinari commented:
@pdf The `Subheader` PR was merged. Could you rebase this PR? It would be great to merge it :+1:.

hindsc52 commented:
Any info on when this will be merged? I could really do with the card title and headers being themeable.

I have tried making the changes manually but to no avail.

pdf commented:
I should be able to take another pass at this on the weekend.

hindsc52 commented:
That would be fantastic, Like I said tried making the changes I can see here by adding the theme colours but that didn't seem to take affect at all. the card background colour changes just not the titles of the subtitles :/

pdf commented:
Well, 334a1a8 and some others make this rebase pretty laborious.  I think it might be easier to rewrite at this stage, will take me a little bit.

pdf commented:
Had to bump eslint-plugin-react to get the docs to build, think this should be pretty close.  There might still be instances of body text color being assumed to be black for the light theme (which can result in unreadable text for dark themes), but I think fixing that for `AppCanvas` and `Paper` should cover most cases.

alitaheri commented:
@pdf This is great! Thanks a lot :+1: :+1: 

I'm all green @oliviertassinari take a look and merge :grin:

hindsc52 commented:
Eagerly awaiting this merge :D thanks for the hardwork @pdf and @alitaheri

alitaheri commented:
Seems like there is a conflict here. @pdf Could you take a look?

pdf commented:
There's a conflict now because a whole truckload of PRs were just merged in front of this again, including stuff like a cleanup of `getStyles` that I mentioned earlier in this very discussion.  Because this PR touches a lot of files, every time this happens, it makes for a painful rebase.  I don't have time to keep fixing this right now.

oliviertassinari commented:
>  I don't have time to keep fixing this right now

Sorry about it! And the migration processes isn't finished yet. We can wait until it's done :grin:.

pdf commented:
Judging by the changes, it'll probably be another re-write for this PR if there're more components to be updated in similar fashion.  I guess just ping me at some stage when you think there's a good time to get this merged.  This is for a personal project though, so I can only work on it in my spare time.

oliviertassinari commented:
> I guess just ping me at some stage when you think there's a good time to get this merged.

Thanks, I will keep you up to date!

newoga commented:
@pdf Would it be alright if I gave you a list of components that are already migrated, then you could start with those?

I really appreciate your patience and work on this.

newoga commented:
Actually on second look the only component you modified that hasn't been migrated yet is `<FloatingActionButton />`, I can finish that now then you should be safe to fix this.

pdf commented:
@newoga rebase conflicts will tell me which components have been updated :wink:, but I won't be able to get back to this for a few days at a minimum (possibly a couple of weeks as I have some other engagements coming up), so don't rush anything on my account.

newoga commented:
Okay no worries then! Thanks much @pdf!

alitaheri commented:
@pdf This requires a lot of work! I'm intercepting this PR. Don't work on it anymore, I'll fix the conflicts tonight so we can merge this soon :+1: Thanks a lot for your initial work on this :+1:

alitaheri commented:
Being continued #3269. 

Thanks @pdf For your hard work on this. You seem busy, so it'll take it from here :+1: :+1: :grin:
